lib/elliptic/eddsa/index.js

```diff
@@ -0,0 +1,143 @@
+'use strict';
+
+var bn = require('bn.js');
+var hash = require('hash.js');
+var elliptic = require('../../elliptic');
+var utils = elliptic.utils;
+var assert = utils.assert;
+var KeyPair = require('./key');
+var Signature = require('./signature');
+
+function parseBytes(bytes) {
+  return typeof bytes === 'string' ? utils.toArray(bytes, 'hex') :
+                                     bytes;
+}
+
+function EDDSA(curve) {
+  if (!(this instanceof EDDSA))
+    return new EDDSA(curve);
+
+  assert(typeof curve === 'string');
+  assert(curve === 'ed25519', 'only tested with ed25519 so far');
+  var curve = this.curve = elliptic.curves[curve].curve;
+  this.g = curve.g;
+  this.g.precompute(curve.n.bitLength() + 1);
+
+  this.pointClass = curve.point().constructor;
+  this.encBytes = Math.ceil(curve.n.bitLength() / 8);
+  this.hash = hash.sha512;
+}
+
+module.exports = EDDSA;
+
+/**
+* @param {Array} message - message bytes
+* @param {Array|String|KeyPair} secret - secret bytes or a keypair
+* @returns {bytes} - signature
+*/
+EDDSA.prototype.sign = function sign(message, secret) {
+  message = parseBytes(message);
+  var key = this.keyFromSecret(secret);
+  var r = this.hashToIntModN(key.messagePrefix(), message);
+  var R = this.g.mul(r);
+  var Rencoded = this.encodePoint(R);
+  var s_ = this.hashToIntModN(Rencoded, key.pubBytes(), message)
+               .mul(key.priv());
+  var S = r.add(s_).mod(this.curve.n);
+  return this.makeSignature({ R: R, S: S, Rencoded: Rencoded });
+};
+
+function pointEq(a, b) {
+  return a.getX().cmp(b.getX()) === 0 &&
+         a.getY().cmp(b.getY()) === 0;
+}
+
+EDDSA.prototype.isPoint = function isPoint(val) {
+  return val instanceof this.pointClass;
+};
+
+/**
+* @param {Array} message - message bytes
+* @param {Array|String|Signature} sig - sig bytes
+* @param {Array|String|Point|KeyPair} pub - public key
+* @returns {bytes} - true if public key matches sig of message
+*/
+EDDSA.prototype.verify = function verify(message, sig, pub) {
+  message = parseBytes(message);
+  sig = this.makeSignature(sig);
+  var key = this.keyFromPublic(pub);
+  var h = this.hashToIntModN(sig.Rencoded(), key.pubBytes(), message);
+  var SG = this.g.mul(sig.S());
+  var RplusAh = sig.R().add(key.pub().mul(h));
+  return pointEq(RplusAh, SG);
+};
+
+/**
+* @returns {bn} little-endian intepretation of hash of one or more
+*               Arrays of bytes.
+*/
+EDDSA.prototype.hashToIntModN = function hashToIntModN() {
```

Comment:
The hash function used, needs to be double bit length of order (or prime I
can't remember). The impls of some of these methods u want in utils will
need parameterising / renaming. Sha512? Little endian? Certain encoding
length. I'd still like to have the convenience methods without needing to
parameterise everywhere. Codec class? Lol